apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/Numeric.cs

```diff
@@ -0,0 +1,663 @@
+            }
+        }
+
+        // Conversions from CLR types to NumericValue (these are as helper for defining operators downstream)
+
+        /// <summary>
+        /// Converts a <see cref="long"/> value to a <see cref="NumericValue"/> with this <see cref="NumericType"/>.
+        /// </summary>
+        /// <param name="value">The integer value to convert.</param>
+        /// <returns>The result of the conversion.</returns>
+        public NumericValue ValueFromLong(long value) => ValueFromScaledInteger(value * PowersOf10[_parameters.Scale]);
+
+        /// <summary>
+        /// Converts a <see cref="ulong"/> value to a <see cref="NumericValue"/> with this <see cref="NumericType"/>.
+        /// </summary>
+        /// <param name="value">The integer value to convert.</param>
+        /// <returns>The result of the conversion.</returns>
+        public NumericValue ValueFromUlong(ulong value) => ValueFromScaledInteger(value * PowersOf10[_parameters.Scale]);
+    }
+
+    internal struct NumericValue : IEquatable<NumericValue>, IComparable<NumericValue>, IComparable
+    {
+        private readonly BigInteger _scaledValue;
+        // Note that this may be null on the case that the NumericValue is initialized to default.
+        // That's fine, we'll check for it, and it's consistent as _scaledValue will be 0, which
+        // is universal, i.e. non-dependent on the actual NumericType, given that the type only varies
+        // in precission and scale.
+        private readonly NumericType _type;
+
+        internal NumericValue(BigInteger scaledValue, NumericType type)
+        {
+            _type = GaxPreconditions.CheckNotNull(type, nameof(type));
+            if (!_type.IsUnscaledValueInRange(scaledValue))
+            {
+                throw new OverflowException("Numeric value out of range");
+            }
+            _scaledValue = scaledValue;
+        }
+
+        internal NumericType Type { get => _type; }
+
+        /// <summary>
+        /// Scales the raw integer value of the value with the smaller scale to the bigger scale.
+        /// </summary>
+        /// <param name="first">The first numeric value to compare scales with.</param>
+        /// <param name="second">The second numeric value to compare scales with.</param>
+        /// <param name="firstScaled">The raw integer value of <paramref name="first"/> possibly scaled to <paramref name="second"/>'s scale
+        /// if <paramref name="second"/>'s scale was the bigger scale.</param>
+        /// <param name="secondScaled">The raw integer value of <paramref name="second"/> possibly scaled to <paramref name="first"/>'s scale
+        /// if <paramref name="first"/>'s scale was the bigger scale.</param>
+        /// <returns>The value of the bigger scale.</returns>
+        private static int GetScaledToBiggerScale(NumericValue first, NumericValue second, out BigInteger firstScaled, out BigInteger secondScaled)
```

Comment:
Do we need this? If we just prohibit comparisons between different types, I suspect we don't.